tests/tests_pytorch/callbacks/test_weight_averaging.py

```diff
@@ -329,3 +329,116 @@ def _train_and_resume(model: TestModel, dataset: Dataset, tmp_path: str, devices
+    callback = EMAWeightAveraging(decay=0.999, update_every_n_steps=1)
+    _train(model, dataset, tmp_path, callback, strategy=strategy, accelerator=accelerator, devices=devices)
+
+    # Verify the average model was created and updated
+    assert callback._average_model is not None
+    assert callback._average_model.n_averaged > 0
+
+
+def test_ema_weight_averaging_step_frequency(tmp_path):
+    """Test EMAWeightAveraging with custom step update frequency."""
+    model = TestModel()
+    dataset = RandomDataset(32, 32)
+
+    # Update every 5 steps
+    callback = EMAWeightAveraging(decay=0.95, update_every_n_steps=5)
+    _train(model, dataset, tmp_path, callback)
+
+    assert callback._average_model is not None
+
+
+def test_ema_weight_averaging_starting_step(tmp_path):
+    """Test EMAWeightAveraging with delayed start based on steps."""
+    model = TestModel()
+    dataset = RandomDataset(32, 32)
+
+    # Start updating after step 10
+    callback = EMAWeightAveraging(decay=0.999, update_every_n_steps=1, update_starting_at_step=10)
+    _train(model, dataset, tmp_path, callback)
+
+    assert callback._average_model is not None
+
+
+def test_ema_weight_averaging_starting_epoch(tmp_path):
+    """Test EMAWeightAveraging with delayed start based on epochs."""
+    model = TestModel()
+    dataset = RandomDataset(32, 32)
+
+    # Start updating after epoch 3
+    callback = EMAWeightAveraging(decay=0.999, update_every_n_steps=1, update_starting_at_epoch=3)
+    _train(model, dataset, tmp_path, callback)
+
+    assert callback._average_model is not None
+
+
+def test_ema_weight_averaging_should_update(tmp_path):
+    """Test the should_update logic of EMAWeightAveraging."""
+    # Test with step-based updates
+    callback = EMAWeightAveraging(update_every_n_steps=5, update_starting_at_step=10)
+
+    # Before starting step
+    assert not callback.should_update(step_idx=5)
+    assert not callback.should_update(step_idx=9)
+
+    # At and after starting step, but not on update frequency
+    assert callback.should_update(step_idx=10)  # First update
+    assert not callback.should_update(step_idx=11)
+    assert not callback.should_update(step_idx=14)
+    assert callback.should_update(step_idx=15)  # Second update
+
+    # Test with epoch-based updates
+    callback = EMAWeightAveraging(update_starting_at_epoch=2)
+
+    assert not callback.should_update(epoch_idx=0)
+    assert not callback.should_update(epoch_idx=1)
+    assert callback.should_update(epoch_idx=2)
+    assert callback.should_update(epoch_idx=3)
+
+
+def test_ema_weight_averaging_checkpoint_save_load(tmp_path):
+    """Test that EMAWeightAveraging correctly saves and loads checkpoints."""
+    model = TestModel()
+    dataset = RandomDataset(32, 32)
+
+    callback = EMAWeightAveraging(decay=0.99, update_every_n_steps=2)
+
+    # Train and create checkpoint
+    _train(model, dataset, tmp_path, callback, will_crash=True)
```

Comment:
Would need to set `crash_on_epoch` for the `TestModel` to raise exception when epoch reaches. Also, this is probably where the tests failed.

```suggestion
    model = TestModel()
    model.crash_on_epoch = 2
    dataset = RandomDataset(32, 32)

    callback = EMAWeightAveraging(decay=0.99, update_every_n_steps=2)

    # Train and create checkpoint
    _train(model, dataset, tmp_path, callback, will_crash=True)
```